### PR TITLE
MOTECH-2504: Improve logging of HTTP-Agent module

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by motech-gerrit** [comment]
Can one of the admins verify this patch?

**Comment by sebbrudzinski** [comment]
test this please

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java [hunk 1]
```diff
@@ -186,7 +186,21 @@ public void handle(MotechEvent motechEvent) {
 
                 LOGGER.info("Posting Http request -- Url: {}, Data: {}",
                         url, String.valueOf(requestData));
-                ResponseEntity<?> response = doExecuteForReturnType(url, requestData, method, restTemplate);
+                ResponseEntity<?> response = null;
+                String errorMessage = null;
+                try {
+                    response = doExecuteForReturnType(url, requestData, method, restTemplate);
+                } catch (Exception e) {
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: Can we just catch HttpException?

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java [hunk 2]
```diff
@@ -186,7 +186,21 @@ public void handle(MotechEvent motechEvent) {
...
+                    errorMessage = e.getMessage();
+                    throw (e instanceof HttpException) ?
+                            (HttpException) e : new HttpException("HTTP error when executing task", e);
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: We don't know if this comes from task execution. Also, there's no need to wrap this exception here. Just rethrow what has been caught.

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java [hunk 3]
```diff
@@ -186,7 +186,21 @@ public void handle(MotechEvent motechEvent) {
...
+                } finally {
+                    if (response != null) {
+                        sendAuditLog(url, requestData, response.getBody().toString(), response.getStatusCode().toString());
+                    } else {
+                        sendAuditLog(url, requestData, errorMessage.substring(4), errorMessage.substring(0, 4));
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: What do we extract here and why?

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java [hunk 4]
```diff
@@ -218,20 +232,26 @@ public void handle(MotechEvent motechEvent) {
...
-        HTTPActionAudit httpActionAudit = new HTTPActionAudit(url, requestData.toString(), response.getBody().toString(),
-                response.getStatusCode().toString());
-        httpActionService.create(httpActionAudit);
+
 
         return response;
     }
 
+    private void sendAuditLog(String url, Object requestData, String body, String statusCode)
+    {
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: This should be in the same line as method signature

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java (hunk 3 above)
Comment (ghost) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: error code and message from server message for database

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java (hunk 1 above)
Comment (ghost) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java]: I tried but got error: exception "HttpException is never thrown in body of corresponding try statment" while compilation

File: http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package org.motechproject.http.agent.handler;
+
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.web.client.ResponseErrorHandler;
+import org.springframework.http.client.ClientHttpResponse;
+import org.motechproject.http.agent.utility.RestUtility;
+
+public class HttpResponseErrorHandler implements ResponseErrorHandler{
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java:13]: Needs javadoc

File: http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpResponseErrorHandler.class);
+
+
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java]: unnecessary new line

File: http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    @Override
+    public void handleError(ClientHttpResponse response) throws IOException {
+        LOGGER.error("Http response error: {} {}", response.getStatusCode(), response.getStatusText());
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/handler/HttpResponseErrorHandler.java]: HTTP

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java [hunk 8]
```diff
@@ -218,18 +225,24 @@ public void handle(MotechEvent motechEvent) {
         } else {
             restTemplate = basicRestTemplate;
         }
+        restTemplate.setErrorHandler(new HttpResponseErrorHandler());
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java:228]: Do we need to set the error handler in two places?

File: http-agent/src/main/java/org/motechproject/http/agent/utility/RestUtility.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package org.motechproject.http.agent.utility;
+
+import org.springframework.http.HttpStatus;
+
```
Comment (sebbrudzinski) [inline http-agent/src/main/java/org/motechproject/http/agent/utility/RestUtility.java:4]: Needs javadoc

**Comment by sebbrudzinski** [comment]
test this please

File: http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java (hunk 8 above)
Comment (ghost) [inline http-agent/src/main/java/org/motechproject/http/agent/listener/HttpClientEventListener.java:228]: i tought that setting error handler for restTemplate should be in same method that restTemplate was created so it's set twice but for two diffrent instances.

**Comment by sebbrudzinski** [comment]
test this please

**Pull request merged**